src/searchbar.cpp

```diff
@@ -113,41 +106,100 @@ void SearchBar::clearSuggestions()
 
 void SearchBar::on_currentTitleChanged(const QString& title)
 {
-    if (this->hasFocus()) {
+    if (title.isEmpty() || title.startsWith("zim://")) {
+        m_title = "";
+        setSearchMode();
+    } else {
+        m_title = title;
+        setBookmarkMode();
+    }
+
+    setText(m_title);
+}
+
+/* We need to intercept mousePressEvent() for cases when QLineEdit
+ * already received the focus before by toggling bookmark,
+ * but now a user clicks into text area of QLineEdit.
+ */
+void SearchBar::mousePressEvent(QMouseEvent *event)
+{
+    // bookmark mode, click either toggles bookmark, or switch into Search state
+    if (event->button() == Qt::LeftButton && isReadOnly()) {
+        qDebug() << "Wait for search action may toggle...";
+        event->accept(); // no process further- don't start editing
+        QTimer::singleShot(50, this, &SearchBar::on_delayed_search_activation);
         return;
     }
-    if (!title.startsWith("zim://")) {
-        setText(title);
-    } else {
-        setText("");
+
+    QLineEdit::mousePressEvent(event);
+}
+
+/* Here we process receiving focus from TAB, or reverse TAB,
+ * or when application window get activated (Alt+TAB).
+ */
+void SearchBar::focusInEvent(QFocusEvent *event)
+{
+    if (event->reason() == Qt::MouseFocusReason) {
+        // mouse click is processed already in mousePressEvent()
+        return;
     }
-    m_button.set_searchMode(title.isEmpty());
-    m_title = title;
+
+    setSearchMode();
 }
 
-void SearchBar::focusInEvent( QFocusEvent* event)
+
+void SearchBar::on_action_triggered()
 {
-    setReadOnly(false);
-    if (event->reason() == Qt::MouseFocusReason && text() == m_title) {
-        clear();
+    auto kiwixApp = KiwixApp::instance();
+    auto library = kiwixApp->getLibrary();
+    auto tabWidget = kiwixApp->getTabWidget();
+
+    if (kiwixApp->isCurrentArticleBookmarked()) {
+        auto zimid = tabWidget->currentZimId();
+        library->removeBookmark(
+            zimid, tabWidget->currentArticleUrl()
+        );
+    } else {
+        kiwix::Bookmark bookmark;
+        auto zimid = tabWidget->currentZimId().toStdString();
+        bookmark.setBookId(zimid);
+        bookmark.setUrl(tabWidget->currentArticleUrl().toStdString());
+        bookmark.setTitle(tabWidget->currentArticleTitle().toStdString());
+        library->addBookmark(bookmark);
     }
-    if (event->reason() == Qt::ActiveWindowFocusReason ||
-        event->reason() == Qt::MouseFocusReason) {
-        connect(&m_completer, QOverload<const QModelIndex &>::of(&QCompleter::activated),
-        this, QOverload<const QModelIndex &>::of(&SearchBar::openCompletion));
+
+    setBookmarkMode(); // already in bookmark mode; called just to set the right (on/off) icon
+    library->save();
+}
+
+void SearchBar::on_delayed_search_activation()
+{
+    if (m_action_toggled) {
```

Comment:
When this is set to true ?

Reply:
Good catch. I have forgotten it somehow, returned back.